### PR TITLE
database moved to postgres

### PR DESCRIPTION
**Pull request by decobots**
#8 #31

### CODE REVIEW

File: src/data_base.py [hunk 1]
```diff
@@ -1,55 +1,71 @@
...
         return self
 
-    def __exit__(self):
-        self.db.close()
-
-    def add_photo(self, id_photo):
-        self.db.execute("INSERT OR IGNORE INTO photos_list(id, posted) VALUES(?,0)", (id_photo,))
-        self.db.commit()
+    def __exit__(self, *args, **kwargs):
+        self.cursor.close()
```
Comment (livercat) [inline src/data_base.py]: Move these two commands to a new .close() method and call it here

File: src/data_base.py [hunk 2]
```diff
@@ -1,55 +1,71 @@
...
-    # def posted_photos(self):
-    #     a = self.db.execute("SELECT id FROM photos_list WHERE posted = 1")
-    #     return [row[0] for row in a]
+    def add_photo(self, id_photo: str):
+        with self.connection as connection:
+            try:
+                connection.cursor().execute(
+                    sql.SQL("INSERT INTO {}(id, posted) VALUES(%s, 'FALSE')").format(sql.Identifier(self.table_name)),
+                    (id_photo,))
+            except psycopg2.IntegrityError:
```
Comment (livercat) [inline src/data_base.py]: See [contextlib.suppress](https://docs.python.org/3/library/contextlib.html#contextlib.suppress)

File: src/data_base.py [hunk 3]
```diff
@@ -1,55 +1,71 @@
...
+            connection.cursor().execute(
+                sql.SQL("UPDATE {} SET posted = 'TRUE' WHERE id = %s").format(sql.Identifier(self.table_name)),
+                (post_id,))
 
-    def delete_photo_from_twitter(self, post_id: str):
-        self.db.execute("UPDATE photos_list SET posted = 0 WHERE id = ?", (post_id,))
-        self.db.commit()
+    def unposted_photos(self) -> List[Photo]:
+        self.cursor.execute(sql.SQL("SELECT id FROM {} WHERE posted = 'FALSE'").format(sql.Identifier(self.table_name)))
+        return [row[0] for row in self.cursor.fetchall()]
```
Comment (livercat) [inline src/data_base.py]: I think you can return generator here, without creating a list

File: tests/test_data_base.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import psycopg2
+import pytest
+from psycopg2 import sql
+
+from src.data_base import DataBase
+
+
+@pytest.fixture
+def table_name():
+    return "test_data_base"
```
Comment (livercat) [inline tests/test_data_base.py]: This should be a constant, not a fixture

File: tests/test_data_base.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+@pytest.fixture
+def db(table_name):
+    database = DataBase(table_name)
+    yield database
+    database.connection.close()
+
+
+@pytest.fixture
+def test_ids():
+    return "42", "43", "44"
```
Comment (livercat) [inline tests/test_data_base.py]: Also a constant

File: tests/test_data_base.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def test_db_context_manager(table_name):
+    with DataBase(table_name) as db:
+        assert db.table_name == table_name
+        assert isinstance(db.connection, psycopg2.extensions.connection)
+        assert isinstance(db.cursor, psycopg2.extensions.cursor)
+        assert db.cursor.closed is False
```
Comment (livercat) [inline tests/test_data_base.py]: assert not db.cursor.closed
Comparing to False or 0 is tying your code to internal psycopg2 implementation

File: tests/test_data_base.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        assert db.connection.closed == 0
```
Comment (livercat) [inline tests/test_data_base.py]: assert not db.connection.closed

File: tests/test_data_base.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    assert db.cursor.closed is True
```
Comment (livercat) [inline tests/test_data_base.py]: assert db.cursor.closed

File: tests/test_data_base.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    assert db.connection.closed == 1
```
Comment (livercat) [inline tests/test_data_base.py]: assert db.connection.closed

File: tests/test_data_base.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def test_add_photo(db, table_name, test_ids):
```
Comment (livercat) [inline tests/test_data_base.py]: Split into test_add_photo and test_add_duplicate. Testing for multiple add's seems redundant.

File: tests/test_data_base.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+def test_post_photo(db, table_name, test_ids):
+    db.post_photo(test_ids[0])
+    db.cursor.execute(sql.SQL("SELECT * FROM {}").format(sql.Identifier(table_name)))
+    list_ids = [(p_id, False) for p_id in test_ids]
+    list_ids[0] = (test_ids[0], True)
+    assert set(db.cursor.fetchall()) == set(list_ids)
+
+
+def test_unposted_photos(db, test_ids):
```
Comment (livercat) [inline tests/test_data_base.py]: Add prerequisites, don't rely on test execution order

File: tests/test_data_base.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def test_unposted_photos_no_left(db, test_ids):
+    for t_id in test_ids:
+        db.post_photo(t_id)
+    unposted = db.unposted_photos()
+    assert unposted == []
+
+
+def teardown_module():
```
Comment (livercat) [inline tests/test_data_base.py]: Move into db fixture

File: src/data_base.py (hunk 1 above)
Comment (decobots) [reply src/data_base.py]: also fixed db fixture in test_data_base

File: tests/test_data_base.py (hunk 4 above)
Comment (decobots) [reply tests/test_data_base.py]: Done

File: tests/test_data_base.py (hunk 5 above)
Comment (decobots) [reply tests/test_data_base.py]: Done

File: src/data_base.py (hunk 2 above)
Comment (decobots) [reply src/data_base.py]: Done

File: src/data_base.py (hunk 3 above)
Comment (decobots) [reply src/data_base.py]: random item could not be chosen from generator

File: tests/test_data_base.py (hunk 6 above)
Comment (decobots) [reply tests/test_data_base.py]: done

File: tests/test_data_base.py (hunk 7 above)
Comment (decobots) [reply tests/test_data_base.py]: done

File: tests/test_data_base.py (hunk 8 above)
Comment (decobots) [reply tests/test_data_base.py]: done

File: tests/test_data_base.py (hunk 9 above)
Comment (decobots) [reply tests/test_data_base.py]: done

File: tests/test_data_base.py (hunk 10 above)
Comment (decobots) [reply tests/test_data_base.py]: done

File: tests/test_data_base.py (hunk 11 above)
Comment (decobots) [reply tests/test_data_base.py]: done, I guess

**Pull request merged**